crowbar_framework/app/models/service_object.rb

```diff
@@ -1478,6 +1463,58 @@ def wait_for_chef_daemons(node_list)
 
   private
 
+  def release_chef_locks(locks)
+    return if locks.empty?
+
+    locks.each_slice(20) do |lock_batch|
+      threads = []
+      lock_batch.each do |lock|
+        threads.push(Thread.new { lock.release })
+      end
+      logger.debug "Waiting for #{threads.length} unlock threads to finish..."
+      ThreadsWait.all_waits(threads)
+    end
+  end
+
+  def lock_and_wait_for_chef_clients(nodes, lock_owner, lock_reason)
+    locks = []
+    errors = {}
+    locks_mutex = Mutex.new
+    errors_mutex = Mutex.new
+
+    nodes.each_slice(20) do |node_batch|
+      threads = {}
+      node_batch.each do |node|
+        thread = Thread.new do
+          begin
+            lock = Crowbar::Lock::SharedNonBlocking.new(
+              logger: @logger,
+              path: "/var/chef/cache/pause-file.lock",
+              node: node,
+              owner: lock_owner,
+              reason: lock_reason
+            ).acquire
+            locks_mutex.synchronize { locks.push(lock) }
+            # Now that we've ensured no new intervallic runs can be started,
+            # wait for any which started before we paused the daemons.
+            wait_for_chef_clients(node, logger: true)
+          rescue Crowbar::Error::LockingFailure => e
+            errors_mutex.synchronize { errors[node] = e.message }
+          end
+        end
+        threads[thread] = node
+      end
+
+      # wait for all running threads and collect the ones with a non-zero return value
+      logger.debug "Waiting for #{threads.keys.length} lock threads to finish..."
+      ThreadsWait.all_waits(threads.keys)
```

Comment:
Doesn't this mean that if just one node in the first batch was super slow for some reason then it would block all subsequent batches?  Or am I reading this wrong?  I would have thought that [something like this](https://blog.engineyard.com/2014/ruby-thread-pool) might be more efficient.